trajopt_sco/src/optimizers.cpp

```diff
@@ -271,11 +222,237 @@ struct MultiCritFilter {
+      std::printf("%15s | %10.3e | %10.3e | %10.3e | %10.3e\n",
+                  cost_names[i].c_str(),
+                  old_cost_vals[i],
+                  approx_improve,
+                  exact_improve,
+                  exact_improve / approx_improve);
+    else
+      std::printf("%15s | %10.3e | %10.3e | %10.3e | %10s\n",
+                  cost_names[i].c_str(),
+                  old_cost_vals[i],
+                  approx_improve,
+                  exact_improve,
+                  "  ------  ");
+  }
+
+  if (cnt_names.size() != 0)
+  {
+    std::printf("%15s | %10s---%10s---%10s---%10s\n",
+                "CONSTRAINTS",
+                "----------",
+                "----------",
+                "----------",
+                "---------"
+                "-");
+    for (size_t i = 0; i < old_cnt_viols.size(); ++i)
+    {
+      double approx_improve = old_cnt_viols[i] - model_cnt_viols[i];
+      double exact_improve = old_cnt_viols[i] - new_cnt_viols[i];
+      if (fabs(approx_improve) > 1e-8)
+        std::printf("%15s | %10.3e | %10.3e | %10.3e | %10.3e\n",
+                    cnt_names[i].c_str(),
+                    merit_error_coeff * old_cnt_viols[i],
+                    merit_error_coeff * approx_improve,
+                    merit_error_coeff * exact_improve,
+                    exact_improve / approx_improve);
+      else
+        std::printf("%15s | %10.3e | %10.3e | %10.3e | %10s\n",
+                    cnt_names[i].c_str(),
+                    merit_error_coeff * old_cnt_viols[i],
+                    merit_error_coeff * approx_improve,
+                    merit_error_coeff * exact_improve,
+                    "  ------  ");
+    }
+  }
+
+  std::printf("%15s | %10.3e | %10.3e | %10.3e | %10.3e\n",
+              "TOTAL",
+              old_merit,
+              approx_merit_improve,
+              exact_merit_improve,
+              merit_improve_ratio);
+}
+
+void BasicTrustRegionSQPResults::writeSolver(std::FILE* stream, bool header) const
+{
+  if (header)
+    std::fprintf(stream, "%s,%s,%s,%s,%s\n", "DESCRIPTION", "oldexact", "dapprox", "dexact", "ratio");
+
+  std::fprintf(stream,
+               "%s,%10.3e,%10.3e,%10.3e,%10.3e\n",
+               "Solver",
+               old_merit,
+               approx_merit_improve,
+               exact_merit_improve,
+               merit_improve_ratio);
+}
+
+void BasicTrustRegionSQPResults::writeVars(std::FILE* stream, bool header) const
+{
+  if (header)
+  {
+    std::fprintf(stream, "%s", "NAMES");
+    for (const auto& var : var_names)
+      std::fprintf(stream, ",%s", var.c_str());
+
+    std::fprintf(stream, "\n");
+  }
+
+  std::fprintf(stream, "%s", "VALUES");
+  for (const auto& x : new_x)
+    std::fprintf(stream, ",%e", x);
+
+  std::fprintf(stream, "\n");
+}
+
+void BasicTrustRegionSQPResults::writeCosts(std::FILE* stream, bool header) const
+{
+  if (header)
+  {
+    std::fprintf(stream, "%s", "COST NAMES");
```

Comment:
what would you think about writing the cost names 4 times each? The reason I say that is that when I pull the resulting log into excel, the costs are above some columns. But right now those aren't the columns they are associated with. If we printed them 4 times each then they would be above the oldexact, dappprox, etc columns that describe that cost.

Same for constraints obviously. I attached some logs for convenience
[trajopt_constraints.log](https://github.com/ros-industrial-consortium/trajopt_ros/files/3611350/trajopt_constraints.log)
[trajopt_costs.log](https://github.com/ros-industrial-consortium/trajopt_ros/files/3611351/trajopt_costs.log)
[trajopt_solver.log](https://github.com/ros-industrial-consortium/trajopt_ros/files/3611352/trajopt_solver.log)
[trajopt_vars.log](https://github.com/ros-industrial-consortium/trajopt_ros/files/3611353/trajopt_vars.log)

Reply:
Good idea. Done.